core/src/main/clojure/xtdb/query.clj

```diff
@@ -136,104 +136,101 @@
+    (when (s/invalid? conformed-query)
+      (throw (err/illegal-arg :malformed-query
+                              {:plan query
+                               :explain (s/explain-data ::lp/logical-plan query)})))
+
+    (let [param-count (or (:param-count (meta query)) 0)
+          param-types-with-defaults (->> (concat
+                                          (mapv #(if (= :default %) :utf8 %) param-types)
+                                          (repeat :utf8))
+                                         (take param-count))
+          tables (filter (comp #{:scan} :op) (lp/child-exprs conformed-query))
+          scan-cols (->> tables
+                         (into #{} (mapcat scan/->scan-cols)))
+
+          _ (assert (or scan-emitter (empty? scan-cols)))
+
+          relevant-schema-at-prepare-time
+          (when (and table-info scan-emitter)
+            (with-open [wm (.openWatermark wm-src)]
+              (->> tables
+                   (map #(str (get-in % [:scan-opts :table])))
+                   (mapcat #(map (partial vector %) (get table-info %)))
+                   (.scanFields scan-emitter wm))))
+
+          cache (ConcurrentHashMap.)
+          ordered-outer-projection (:named-projection (meta query))
+          param-fields (mapify-params (mapv (comp types/col-type->field types/col-type->nullable-col-type) param-types-with-defaults))
+          default-tz (or default-tz (.getZone expr/*clock*))]
+
+      (reify PreparedQuery
+        (paramFields [_] param-fields)
+        (columnFields [_]
+          (let [{:keys [fields]} (emit-expr cache deps conformed-query scan-cols default-tz (into {} param-fields))]
+            ;; could store column-fields in the cache/map too
+            (->column-fields ordered-outer-projection fields)))
+        (bind [_ {:keys [args params basis default-tz]
+                  :or {default-tz default-tz}}]
+
+          ;; TODO throw if basis is in the future?
+          (util/with-close-on-catch [args (open-args allocator args)]
+            ;;TODO consider making the either/or relationship between params/args explicit, e.g throw error if both are provided
+            (let [params (or params args)
+                  {:keys [fields ->cursor]} (emit-expr cache deps conformed-query scan-cols default-tz (->param-fields params))
+                  {:keys [current-time]} basis
+                  current-time (or current-time (.instant expr/*clock*))
+                  clock (Clock/fixed current-time default-tz)]
+
+              (reify
+                BoundQuery
+                (columnFields [_]
+                  (->column-fields ordered-outer-projection fields))
+                (openCursor [_]
+                  (when relevant-schema-at-prepare-time
+                    (let [table-info-at-execution-time (with-open [wm (.openWatermark wm-src)]
+                                                         (.scanFields scan-emitter wm
```

Comment:
I'm assuming this is next to go onto the wm?

Reply:
Yes, I think it should be possible to retire scanFields entirely, replacing it with a schema-map with Fields as leaf entries.